### ISSUE TITLE
Unsigned integer outputs become signed

### ISSUE THREAD

**Issue by hawkinsp** [openxla/iree#9282]
Repro:
```
import platform
import iree.compiler
import iree.runtime
import numpy as np

CODE = """
#loc0 = loc(unknown)
module @jit_prim_fun.0 {
  func.func public @main(%arg0: tensor<5x7xui32> loc(unknown)) -> tensor<5x7xui32> {
    %0 = "mhlo.sort"(%arg0) ({
    ^bb0(%arg1: tensor<ui32> loc(unknown), %arg2: tensor<ui32> loc(unknown)):
      %1 = "mhlo.compare"(%arg1, %arg2) {compare_type = #mhlo<"comparison_type UNSIGNED">, comparison_direction = #mhlo<"comparison_direction LT">} : (tensor<ui32>, tensor<ui32>) -> tensor<i1> loc(#loc2)
      "mhlo.return"(%1) : (tensor<i1>) -> () loc(#loc1)
    }) {dimension = 1 : i64, is_stable = false} : (tensor<5x7xui32>) -> tensor<5x7xui32> loc(#loc1)
    return %0 : tensor<5x7xui32> loc(#loc0)
  } loc(#loc0)
} loc(#loc0)
#loc1 = loc("jit(sort)/jit(main)/sort[dimension=1 is_stable=False num_keys=1]"("/Users/phawkins/p/jax/tests/lax_test.py":2039:1))
#loc2 = loc("jit(sort)/jit(main)/lt"("/Users/phawkins/p/jax/tests/lax_test.py":2039:1))
"""

# Work around https://github.com/google/iree/issues/9028
extra_args = []
if platform.system() == "Darwin" and platform.machine() == "arm64":
  extra_args += ["--iree-llvm-target-triple=arm64-apple-darwin21.5.0"]

iree_config = iree.runtime.system_api.Config("dylib")
iree_binary = iree.compiler.compile_str(
    CODE, target_backends=["dylib"], input_type="mhlo", extra_args=extra_args)
vm_module = iree.runtime.VmModule.from_flatbuffer(iree_binary)
module_object = iree.runtime.load_vm_module(vm_module, iree_config)

out = module_object["main"](np.ones((5, 7), dtype=np.uint32))
assert out.dtype == np.uint32, out.dtype
```

Get:
```
Traceback (most recent call last):
  File "/Users/phawkins/p/jax/t.py", line 34, in <module>
    assert out.dtype == np.uint32, out.dtype
AssertionError: int32
```

The MHLO has a ui32 output, so I expect a uint32 result from IREE.

**Comment by trevor-m** [comment]
I took a look into this bug and it seems the type signedness is intentionally discarded during the `MHLOToLinalgOnTensors` pass. This is done because the LLVMs dialect such as linalg, arith, etc uses the operators to convey signedness (ex. `arith.cmpi slt`/`arith.cmpi ult`), while MHLO uses the type (`i32` /`ui32`).

```
// -----// IR Dump After ConvertMHLOToLinalgOnTensors (iree-mhlo-to-linalg-on-tensors) //----- //
#loc = loc(unknown)
module @jit_prim_fun {
  func.func public @main(%arg0: tensor<5x7xi32> loc(unknown)) -> tensor<5x7xi32> {
    %0 = iree_linalg_ext.sort dimension(1) outs(%arg0 : tensor<5x7xi32>) {
    ^bb0(%arg1: i32 loc(unknown), %arg2: i32 loc(unknown)):
      %1 = arith.cmpi ult, %arg1, %arg2 : i32 loc(#loc3)
      iree_linalg_ext.yield %1 : i1 loc(#loc2)
    } -> tensor<5x7xi32> loc(#loc2)
    return %0 : tensor<5x7xi32> loc(#loc)
  } loc(#loc)
} loc(#loc)
#loc1 = loc("/Users/phawkins/p/jax/tests/lax_test.py":2039:1)
#loc2 = loc("jit(sort)/jit(main)/sort[dimension=1 is_stable=False num_keys=1]"(#loc1))
#loc3 = loc("jit(sort)/jit(main)/lt"(#loc1))
```

It seems that we need to retain the orginal type somewhere in the IR in order to fix this. I'm not sure what is the best way to accomplish that. Maybe it could be done using some kind of attribute in the outer module or with a cast op on the output? Or should the MLIR dialects retain the signedness in the type and just ignore it while deferring to the signedness in the operations?

**Comment by stellaraccident** [comment]
We need to isolate which layer is breaking down here. Downward, below MHLO, types should be fully translated to signless (where the operations carry sign information where needed).

I expect that this is just at our Python layer taking the signless type in the signature and mapping that to a signed dtype when producing the array. If this is true, bitcasting to the expected unsigned dtype should show that the computation was done properly.

The last step of converting an IREE internal ElementType to dtype happens here: https://github.com/openxla/iree/blob/main/runtime/bindings/python/hal.cc#L439

I expect we will find that this is decoding IREE_HAL_ELEMENT_TYPE_INT_32 (signless int32) as a signed int32 and that we are not setting the ElementType to a fully signed dtype in the compiler. I'd need to dig into where we do that but would be good to know if the hypothesis holds.

**Comment by trevor-m** [comment]
Hi @stellaraccident, thanks for the response and suggestions!
Your hypothesis about the python mappings looks to be correct. The element type is coming in as `IREE_HAL_ELEMENT_TYPE_INT_32` which is mapped to `type_code = i`. If I hardcode the `type_code = I`, the output type is now correctly `uint32` when I run the script.  In either case, the computation is correct if the output is bitcasted.

**Comment by jpienaar** [comment]
I believe Rob is looking at this one actively.

**Comment by rsuderman** [comment]
I added additional details to the duplicate that exists in the JAX integration work: https://github.com/openxla/iree/issues/12665

**Comment by julianwa** [comment]
@rsuderman should this be closed as a duplicate of #12665 ?

**Comment by rsuderman** [comment]
> @rsuderman should this be closed as a duplicate of #12665 ?

Probably, I usually just link all of the fixes together and close as a group. Sometimes people see a bug closes and assume its because it was fixed (not that it was a dupe).
